docs/compiler-guide.md

```diff
@@ -0,0 +1,565 @@
+
+In contract, FCS supports compiling a graph of projects, each for a different assembly. The Typed Tree nodes are **not** shared between different project compilations. This means that representation of project references is roughly O(n^2) in memory usage. In practice it's not strictly O(n^2), but for very large codebases the proportionality is felt.
+
+Some key things to understand are:
+
+* The `RawFSharpAssemblyData` is the data blob that would normally be stuffed in the F# resource in the generated DLL  in a normal compilation. That's the "output" of checking each project.
+
+* This is used as "input" for the assembly reference of each consuming project (instead of an on-disk DLL)
+
+* Within each consuming project that blob is then resurrected to Typed Tree nodes in `TypedTreePickle.fs`.
+
+The biggest question is: could the compiler share this data across projects? In theory, yes. In practice, it's very tricky business.
+
+From a correctness point of view: the process of generating this blob (TypedTreePickle `p_XYZ`) and resurrecting it (TypedTreePickle `u_*`) does some transformations to the Typed Tree that are necessary for correctness of compilation, for example, [in `TypedTreePickle`](https://github.com/dotnet/fsharp/blob/master/src/fsharp/TypedTreePickle.fs#L737). Basically, the Typed Tree nodes from the compilation of one assembly are _not_ valid when compiling a different assembly.
+
+The Typed Tree nodes include `CcuData` nodes, which have access to a number of callbacks into the `TcImports` compilation context for the assembly being compiled. TypedTree nodes are effectively tied to a particular compilation of a particular assembly due to this.
+
+There isn't any way to share this data without losing correctness and invalidating many invariants held in the current design.
+
+From a lifetime point of view: the Typed Tree nodes are tied together in a graph, so sharing one or two of them might drag across the entire graph and extend lifetimes of that graph. None of these interrelated nodes were designed to be shared across assemblies.
+
+## `eventually` computations
+
+Some parts of the F# codebase (specifically, the type checker) are written using `eventually` computation expressions. These define resumption-like computations which can be  time-sliced, suspended or discarded at "bind" points.
+
+This is done to ensure that long-running type-checking and other computations in the F# Compiler Service can be interrupted and cancelled. The documentation of the [F# Compiler Service Operations Queue](fsharp-compiler-service-queue.md) covers some aspects of this.
+
+When compiling code with `fsc` or executing code with `fsi`, these computations are not time-sliced and simply run synchronously and without interruption (unless the user requests it).
+
+TODO for don --> is the below stuff still accurate?
```

Comment:
@dsyme I need your help checking if the words in the rest of this section are still accurate

Reply:
I think we should integrate now and then later today I do an end to end edit as a separate PR

Reply:
Okay. Can you approve?